### DISCUSSION TITLE
Calling rust function from slint

### DISCUSSION THREAD

**Discussion by realroot2185** [slint-ui/slint#7083] [Q&A] [answered]
In appwindow.slint I have a component from another slint file which again as a component from another slint file.
In the latter when a Timer ends I like to call a function in the rust backend.
How can I do that?

Since the component is not the main I am not sure to how send it to appwindow.slint and use some handler.

Not sure what's the best method here.

**Comment by ultimaweapon** [comment]
What I did is forward the callback to the root component. I think another solution is using a [global singleton](https://docs.slint.dev/latest/docs/slint/src/language/syntax/globals).

**Reply by realroot2185** [reply]
>What I did is forward the callback to the root component.

How?

I made a global singleton in appwindow.slint when I am in that file I can access it but elsewhere I cannot.

Do I have to define all together?

**Reply by Vadoola** [reply]
> How?
> 
> I made a global singleton in appwindow.slint when I am in that file I can access it but elsewhere I cannot.


Did you import it into your other component?

**Reply by realroot2185** [reply]
appwindow.slint imports a component (to use it):
`import { Timers } from "timers.slint";`

That imports another component:
`import { MyTimer } from "my_timer.slint";`

Which imports Logic:
 `import { Logic } from "appwindow.slint";`:

For some reason:
```
 error: Recursive import of "/path/app/ui/timers.slint"
   --> /path/app/ui/appwindow.slint:4:24
    |
  4 | import { Timers } from "timers.slint";
```

Not sure what I am missing.

**Reply by Vadoola** [reply]
I've never seen this error, but I could see it. Your app window imports, timer which imports my timer which imports logic...which I'm guess needs timer and you go around in a circle.

I would probably move your global Logic into it its own Slint file, that's how I've done it instinctually and it's never caused me an issue. There certainly could be other ways around it though. But yeah I feel like move your global Logic into a Logic.slint and import it elsewhere where needed.

**Reply by realroot2185** [reply]
How do connect it in main.rs in that case?
I have `import { Logic } from "global.slint"` in appwindow.slint and my_timer.slint.
```
//main.rs
 let app = App::new()?;
    app.global::<Logic>().on_completed(|| {
    [...]

error[E0412]: cannot find type `Logic` in this scope
  --> src/main.rs:17:18
   |
17 |     app.global::<Logic>().on_completed(|| {
   |                  ^^^^^ not found in this scope
```

If I declare `Logic {}` in appwindow.slint it does not compile: `error: Cannot create an instance of a global component`.

**Reply by Vadoola** [reply]
You need to re-export it from Appwindow.slint.

I'm on mobile at the moment but it should be something like

`import { Logic } from "global.slint"

Export Logic;
`

If I have the syntax slightly wrong there should be some examples elsewhere in the Slint discussion, or look at the Main Window of my app Tomotroid. There should also be something in the Slint docs about re-exporting.

**Reply by realroot2185** [reply]
```
export Logic;

error: Parse error: expected a top-level item such as a component, a struct, or a global
```
Nice a slint timer app.
It's not so easy to see how because it is not so simple as what I have here.
I think that I will look into that thanks.

https://github.com/slint-ui/slint/discussions/6757
Maybe I need to define my_timer.slint in appwindow.slint because the docs and other discussion have all in one file.

**Reply by Vadoola** [reply, marked as answer]
You shouldn't need to put it all in one module here is the docs that includes re-exporting: https://releases.slint.dev/1.0.1/docs/slint/src/reference/modules

It looks like in my example above I forgot to out braces around Logic.

It should be

`import { Logic } from "global.slint"

Export {Logic}`
